### PR TITLE
Transaction-ID removal

### LINKED ISSUE #25: transaction-id shortcoming related to accounts
the uniqueness property of utxo is hard to satisfy when the component are just accounts, since the nonce is associated with the witness only.

### PR DESCRIPTION
**Pull request by vincenthz**
remove transaction-id for transaction, and replace by:

* `TransactionSignDataHash` for the data that need to be signed for authenticating the transaction
* `FragmentId` which is used to store the transaction and identify it uniquely in the ledger

fix #25

### CODE REVIEW

File: chain-impl-mockchain/src/ledger/tests/discrimination_tests.rs [hunk 1]
```diff
@@ -9,6 +10,7 @@ use crate::testing::{
 use crate::transaction::*;
 use crate::value::*;
 use chain_addr::Discrimination;
+use chain_core::property::Message as _;
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/ledger/tests/discrimination_tests.rs]: I'll start renaming the network things as well, in a separate PR.

File: chain-impl-mockchain/src/transaction/transaction.rs [hunk 2]
```diff
@@ -1,12 +1,26 @@
...
 use crate::value::{Value, ValueError};
 use chain_addr::Address;
 use chain_core::mempack::{read_vec, ReadBuf, ReadError, Readable};
 use chain_core::property;
+use chain_crypto::{digest::DigestOf, Blake2b256};
+use std::boxed::Box;
 
-// FIXME: should this be a wrapper type?
-pub type TransactionId = Hash;
+pub struct TransactionSignData(Box<[u8]>);
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/transaction/transaction.rs]: Good, a frozen `Vec`.

File: chain-impl-mockchain/src/transaction/transaction.rs [hunk 3]
```diff
@@ -112,10 +126,10 @@ impl<Extra: property::Serialize> Transaction<Address, Extra> {
         self.serialize_body(&mut codec.into_inner())
     }
 
-    pub fn hash(&self) -> TransactionId {
+    pub fn hash(&self) -> TransactionSignDataHash {
         let mut bytes = Vec::new();
         self.serialize_body(&mut bytes).unwrap();
-        TransactionId::hash_bytes(&bytes)
+        DigestOf::digest(&TransactionSignData(bytes.into()))
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/transaction/transaction.rs]: I think there should be a better looking API, like `Digest::of` to produce `DigestOf`.
Something to think of for a future refactoring.

File: chain-impl-mockchain/src/ledger/tests/discrimination_tests.rs (hunk 1 above)
Comment (mzabaluev) [reply chain-impl-mockchain/src/ledger/tests/discrimination_tests.rs]: See input-output-hk/jormungandr#523

File: chain-impl-mockchain/src/transaction/transaction.rs (hunk 3 above)
Comment (vincenthz) [reply chain-impl-mockchain/src/transaction/transaction.rs]: ah that's a good name, will add it

**Pull request merged**